File: VB/WindowsApplication53/PivotHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraPivotGrid
Imports System.IO
Imports System.Runtime.Serialization.Formatters.Binary

Namespace WindowsApplication53
	Public Class PivotLayoutHelper
		Public Shared Sub SavePivot(ByVal pivot As PivotGridControl, ByVal stream As Stream)
			Dim s As New Storage()
			Using layoutStream As New MemoryStream()
				pivot.SaveLayoutToStream(layoutStream, PivotGridOptionsLayout.FullLayout)
				s.layout = layoutStream.ToArray()
			End Using
			Using stateStream As New MemoryStream()
				pivot.SaveCollapsedStateToStream(stateStream)
				s.collapsedState = stateStream.ToArray()
			End Using
			Dim binFormat As New BinaryFormatter()
			binFormat.Serialize(stream, s)

		End Sub
		Public Shared Sub LoadPivot(ByVal pivot As PivotGridControl, ByVal stream As Stream)
			Dim s As New Storage()

			Dim binFormat As New BinaryFormatter()
			s = TryCast(binFormat.Deserialize(stream), Storage)

			Using layoutStream As New MemoryStream()
				layoutStream.Write(s.layout, 0, s.layout.Length)
				layoutStream.Position = 0
				pivot.RestoreLayoutFromStream(layoutStream, PivotGridOptionsLayout.FullLayout)

			End Using
			Using stateStream As New MemoryStream()
				stateStream.Write(s.collapsedState, 0, s.collapsedState.Length)
				stateStream.Position = 0
				pivot.LoadCollapsedStateFromStream(stateStream)
			End Using

		End Sub

		<Serializable> _
		Public Class Storage
			Public layout() As Byte
			Public collapsedState() As Byte
		End Class
	End Class
End Namespace

File: CS/WindowsApplication53/PivotHelper.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraPivotGrid;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;

namespace WindowsApplication53
{
    public class PivotLayoutHelper
    {
        public static void SavePivot(PivotGridControl pivot, Stream stream)
        {
            Storage s = new Storage();
            using (MemoryStream layoutStream = new MemoryStream())
            {
                pivot.SaveLayoutToStream(layoutStream, OptionsLayoutPivotGrid.FullLayout);
                s.layout = layoutStream.ToArray();
            }
            using (MemoryStream stateStream = new MemoryStream())
            {
                pivot.SaveCollapsedStateToStream(stateStream);
                s.collapsedState = stateStream.ToArray();
            }
            BinaryFormatter binFormat = new BinaryFormatter();
            binFormat.Serialize(stream, s);

        }
        public static void LoadPivot(PivotGridControl pivot, Stream stream)
        {
            Storage s = new Storage();

            BinaryFormatter binFormat = new BinaryFormatter();
            s = binFormat.Deserialize(stream) as Storage;

            using (MemoryStream layoutStream = new MemoryStream())
            {
                layoutStream.Write(s.layout, 0, s.layout.Length);
                layoutStream.Position = 0;
                pivot.RestoreLayoutFromStream(layoutStream, OptionsLayoutPivotGrid.FullLayout);

            }
            using (MemoryStream stateStream = new MemoryStream())
            {
                stateStream.Write(s.collapsedState, 0, s.collapsedState.Length);
                stateStream.Position = 0;
                pivot.LoadCollapsedStateFromStream(stateStream);
            }

        }

        [Serializable]
        public class Storage
        {
            public byte[] layout;
            public byte[] collapsedState;
        }
    }
}
